async-nats/src/lib.rs

```diff
@@ -521,21 +454,104 @@ impl SubscriptionContext {
+///
+pub struct Connector {
+    server_addrs: Vec<ServerAddr>,
+    options: ConnectOptions,
+}
+
+impl Connector {
+    pub(crate) async fn connect(&mut self) -> Result<(ServerInfo, Connection), io::Error> {
+        loop {
+            for server_addr in &self.server_addrs {
+                if let Ok((server_info, connection)) = self.connect_to(server_addr).await {
+                    return Ok((server_info, connection));
+                }
+            }
+        }
+    }
+
+    pub(crate) async fn connect_to(
+        &self,
+        server_addr: &ServerAddr,
+    ) -> Result<(ServerInfo, Connection), io::Error> {
+        let tls_config = tls::config_tls(&self.options).await?;
+
+        let tcp_stream = TcpStream::connect((server_addr.host(), server_addr.port())).await?;
+        tcp_stream.set_nodelay(true)?;
+
+        let mut connection = Connection {
+            stream: Box::new(BufWriter::new(tcp_stream)),
+            buffer: BytesMut::new(),
+        };
+
+        let op = connection.read_op().await?;
+        let info = match op {
+            Some(ServerOp::Info(info)) => info,
+            Some(op) => {
+                return Err(io::Error::new(
+                    ErrorKind::Other,
+                    format!("expected INFO, got {:?}", op),
+                ))
+            }
+            None => {
+                return Err(io::Error::new(
+                    ErrorKind::Other,
+                    "expected INFO, got nothing",
+                ))
+            }
+        };
+
+        let tls_required =
+            self.options.tls_required || info.tls_required || server_addr.tls_required();
+
+        if tls_required {
+            let tls_config = Arc::new(tls_config);
+            let tls_connector =
+                tokio_rustls::TlsConnector::try_from(tls_config).map_err(|err| {
+                    io::Error::new(
+                        ErrorKind::Other,
+                        format!("failed to create TLS connector from TLS config: {}", err),
+                    )
+                })?;
+
+            let domain = rustls::ServerName::try_from(info.host.as_str())
+                .or_else(|_| rustls::ServerName::try_from(server_addr.host()))
+                .map_err(|_| {
+                    io::Error::new(
+                        ErrorKind::InvalidInput,
+                        "cannot determine hostname for TLS connection",
+                    )
+                })?;
+
+            connection = Connection {
+                stream: Box::new(tls_connector.connect(domain, connection.stream).await?),
+                buffer: BytesMut::new(),
+            };
+        };
+
+        Ok((*info, connection))
+    }
+}
+
+/// A handler which facilitates communication from channels to a single shared connection.
+/// The handler takes ownership of the channel.
 ///
 /// The type will probably not be public.
-pub(crate) struct Connector {
+pub(crate) struct Handler {
```

Comment:
I don't like the name, but I don't have a better one yet.

I see `Handler` that handles some request/op that came and then ends.
Here we have an ongoing process. It also tell nothing about what it does (would be fine if we had it in `connect` or `connection` module, but we don't, at least as of yet).

Maybe `ConnectionHandler`

Reply:
Yeah this was a case of "oh I need connector to be an actual connector" 😉